### ISSUE TITLE
Regression deploying AWS apps as native image with 2.3.0

### ISSUE THREAD

**Issue by ilopmar** [micronaut-projects/micronaut-core#4852]
I've discovered this issue this morning when finishing the update of the GraalVM AWS test applications. This was working last Friday.

When deploying an AWS app as GraalVM and you send a request, it fails with:

```
10:14:15.448 [main] INFO  io.micronaut.runtime.Micronaut - No embedded container found. Running as CLI application
time="2021-01-25T10:14:15.45" level=error msg="Init failed" InvokeID= error="Runtime exited without providing a reason"
time="2021-01-25T10:14:15.45" level=error msg="INIT DONE failed: Runtime.ExitError"
Invalid lambda response received: Lambda response must be valid json
2021-01-25 11:14:15 127.0.0.1 - - [25/Jan/2021 11:14:15] "GET /jokes/nerdy HTTP/1.1" 502 
```

### Steps to Reproduce

- `git clone https://github.com/micronaut-graal-tests/micronaut-aws-app-graal`
- `cd micronaut-aws-app-graal`
- `git checkout 2.3.x`
- `sdk use java 21.0.0.r8-grl`
- `./build-native-image.sh`

That will create the file `build/function.zip`. Now we can test it with SAM local:

```
./sam-local.sh
```

Send a request:

```
curl localhost:3000/jokes/nerdy
```

### Expected Behaviour

The application show response properly.

### Actual Behaviour

It fails with `{"message":"Internal server error"}` and in the SAM local console:

```
$ ./sam-local.sh                                                                 130 ↵
Mounting GraalVMApiService at http://127.0.0.1:3000/{proxy+} [DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT]
You can now browse to the above endpoints to invoke your functions. You do not need to restart/reload SAM CLI while working on your functions, changes will be reflected instantly/automatically. You only need to restart SAM CLI if you update your AWS SAM template
2021-01-25 11:14:11  * Running on http://127.0.0.1:3000/ (Press CTRL+C to quit)
Invoking not.used.in.provided.runtime (provided)
Decompressing /home/ivan/workspaces/micronaut-graal-tests/micronaut-aws-app-graal/build/function.zip
Skip pulling image and use local one: amazon/aws-sam-cli-emulation-image-provided:rapid-1.15.0.

Mounting /tmp/tmprofr6gnm as /var/task:ro,delegated inside runtime container
START RequestId: 59059557-4887-4a02-8738-519ea2fd402d Version: $LATEST
 __  __ _                                  _   
|  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_ 
| |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
| |  | | | (__| | | (_) | | | | (_| | |_| | |_ 
|_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
  Micronaut (v2.3.0-SNAPSHOT)

10:14:15.448 [main] INFO  io.micronaut.runtime.Micronaut - No embedded container found. Running as CLI application
time="2021-01-25T10:14:15.45" level=error msg="Init failed" InvokeID= error="Runtime exited without providing a reason"
time="2021-01-25T10:14:15.45" level=error msg="INIT DONE failed: Runtime.ExitError"
Invalid lambda response received: Lambda response must be valid json
2021-01-25 11:14:15 127.0.0.1 - - [25/Jan/2021 11:14:15] "GET /jokes/nerdy HTTP/1.1" 502 -
```

Micronaut AWS has been upgraded for Micronaut 2.3.0. I've also tried with Micronaut AWS 2.2.5 but it fails with the same error, so probably is something done in micronaut-core, but not sure what.
The same error happens when deploying the application directly to AWS (you can do it following the steps in the README.md file).

### Environment Information

- **Operating System**: Linux Mint 20
- **Micronaut Version:** 2.3.0
- **JDK Version:** 8

**Comment by graemerocher** [comment]
From the output it looks like the `Application` class is being run instead of the Lambda request handler

**Comment by graemerocher** [comment]
For Lambda the `mainClass` has to be `io.micronaut.function.aws.runtime.MicronautLambdaRuntime` or some other runtime impl. Something has changed in either starter or the Gradle plugin where this is now set to the `Application` class instead

**Comment by ilopmar** [comment]
Thank you for the hint Graeme, that was exactly what was happening and I now see why it worked on Friday and fails now.
I was using `buildNativeLambda` task on Friday that takes care of everything and it adds the right main class `-H:Class=io.micronaut.function.aws.runtime.MicronautLambdaRuntime`.

This morning I've changed that because I can't use the `buildNativeLambda` task in CI. This is because I wan't to use the GraalVM distribution compiled from source code and not the stable one that the previous task uses. That's why the main class wasn't correct.

In any case, with the "manual" approach it still fails. I've setup the right main class `mainClass.set("io.micronaut.function.aws.runtime.MicronautLambdaRuntime")` and added the following to build.gradle:

```
nativeImage {
    args = ['--verbose']
}
```

Then I run `./gradlew nativeImage` and I see that the correct main class is used:

```
...
-H:Class=io.micronaut.function.aws.runtime.MicronautLambdaRuntime \
...
```

I package the native-image and the bootstrap file I create manually but it still fails with the same error.

**Comment by graemerocher** [comment]
not sure what you're doing then because I just created a new 2.3.0 app switched the main class to `io.micronaut.function.aws.runtime.MicronautLambdaRuntime` and deployed and it worked fine

**Comment by ilopmar** [comment]
I'm doing the same and it still fails:

- `mn create-app test-aws --features=aws-lambda --jdk 8`
- Replace the `mainClass`
- Modify generated `BookController` and add `@Controller("/books")`
- `./gradlew nativeImage`
- Create `bootstrap`:

```
#!/bin/sh
set -euo pipefail
./application
```

- Package the function.zip: `zip -j build/function.zip bootstrap build/native-image/application`

- Create the file `sam-native.yaml` with:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Description: AWS Serverless Micronaut API
Globals:
  Api:
    EndpointConfiguration: REGIONAL
Resources:
  GraalVMApiService:
    Type: AWS::Serverless::Function
    Properties:
      Handler: not.used.in.provided.runtime
      Runtime: provided
      CodeUri: build/function.zip
      MemorySize: 512
      Policies: AWSLambdaBasicExecutionRole
      Tracing: Active
      Timeout: 15
      Events:
        GetResource:
          Type: Api
          Properties:
            Path: /{proxy+}
            Method: any
  LogsLogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      RetentionInDays: 7

Outputs:
  MicronautNativeServiceApi:
    Description: URL for application
    Value: !Sub 'https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod'
    Export:
      Name: MicronautNativeServiceApi
```

- Start it: `sam local start-api -t sam-native.yaml`
- Send request: `curl -X POST -H 'Content-Type:application/json' -d '{"name":"My book"}' localhost:3000/books`


If I run `./gradlew buildNativeLambda && cp build/libs/test-aws-0.1-lambda.zip build/function.zip` and run `sam local ...` again and send the same curl request it works.

Can you please double-check with this steps or share the app you are testing with?

**Comment by graemerocher** [comment]
Everything you did except I deployed to AWS not via SAM

**Comment by graemerocher** [comment]
but yes that is expected. I wouldn't expect `nativeImage` to work by itself

**Comment by ilopmar** [comment]
Why do you expect `nativeImage` won't work? I mean it generates the native image with the right mainClass. Then I should be able to manually zip it with bootstrap and upload to lambda, right?

**Comment by graemerocher** [comment]
currently `buildNativeLambda` is the only supported way to build a native image for a lambda that I am aware of.

If you are doing it your own way and rolling your own then you are on your own :)

If you wish to look at getting `nativeImage` working like you want it to then the changes will need to happen in the Gradle plugin, where PRs are of course welcome

**Comment by graemerocher** [comment]
One of the reasons I expect `nativeImage` will not work is because you need to build a Lambda with Amazon linux and since `nativeImage` doesn't run within a docker container that is not possible. 

Of course if you want to run Amazon Linux as your operating system so you can build native images with `nativeImage` then that is up to you.

**Comment by ilopmar** [comment]
The main issue on my side was that Linux Brew (which I used for aws/sam CLIs) were corrupted and that generated a wrong image somehow. I've got rid of it and installed the official `aws` client and now deployed apps work. I'll try it again later on sam local (but that has worked before)

Regarding the `nativeImage` task, it also works now doing:

```
./gradlew nativeImage
zip -j build/function.zip bootstrap build/native-image/application
```

And then deploying `build/function.zip` to AWS Lambda.

I don't need to use Amazon Linux and we haven't needed it in the past. For example, in CI we've been running `native-image` command and building the `function.zip` manually for a while without using Amazon Linux. I agree that using it is the best approach because we make sure that we build the image with the same Linux distro that is used in AWS, but for CI should be fine using other Linux.

I'm closing this and tagging it as "not a bug" because I think everything will work now once I modify the build a little bit.
